topology/probes/bess/bess.go

```diff
@@ -0,0 +1,248 @@
+	}
+	for _, oGate := range modRes.GetOgates() {
+		p.walkAndRender(oGate.GetName(), parentNode, edgeList)
+		edgeID := graph.GenID(moduleName, string(oGate.Igate), string(oGate.Ogate), oGate.GetName())
+		if p.Ctx.Graph.GetEdge(edgeID) == nil {
+			p.Ctx.Graph.NewEdge(edgeID, node,
+				p.Ctx.Graph.GetNode(graph.GenID(oGate.GetName())),
+				graph.Metadata{"RelationType": "layer2", "Directed": true, "Backend": BessBackend})
+		}
+		*edgeList = append(*edgeList, p.Ctx.Graph.GetEdge(edgeID))
+	}
+
+}
+
+func (p *Probe) Do(ctx context.Context, wg *sync.WaitGroup) error {
+	p.Ctx.Logger.Info("Connecting to BESS")
+	conn, err := grpc.Dial(fmt.Sprintf("%s:%d", p.bessHost, p.bessPort), grpc.WithInsecure())
+	if err != nil {
+		return err
+	}
+	p.Ctx.Logger.Info("Connected to BESS")
+	p.bessClient = bess_pb.NewBESSControlClient(conn)
+	verResponse, err := p.bessClient.GetVersion(context.Background(), &bess_pb.EmptyRequest{})
+	if err != nil || verResponse.Error != nil {
+		return errors.New(fmt.Sprintf("Could not get version: %s, %s", verResponse.GetError(), err))
+	}
+	p.Ctx.Logger.Infof("BESS connected with version: %s", verResponse.Version)
+
+	metadata := graph.Metadata{
+		"Name":    "bess",
+		"Type":    "bess",
+		"Program": "bessd",
+		"Version": verResponse.GetVersion(),
+		"Backend": BessBackend,
+	}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		// probe running function
+		for {
+			// get modules and data from bess
+			modsRes, err := p.bessClient.ListModules(p.bessContext, &bess_pb.EmptyRequest{})
+			if err != nil {
+				p.Ctx.Logger.Errorf("Unable to query module list for BESS: %s", err)
+				return
+			}
+
+			portRes, err := p.bessClient.ListPorts(p.bessContext, &bess_pb.EmptyRequest{})
+			if err != nil {
+				p.Ctx.Logger.Errorf("Unable to query port list for BESS: %s", err)
+				return
+			}
+
+			var foundMods []string
+			var foundEdges []*graph.Edge
+			p.Ctx.Graph.Lock()
+
+			// build main bess node
+			bessID := graph.GenID("bessd", p.bessHost)
+			bessNode := p.Ctx.Graph.GetNode(bessID)
+			if bessNode == nil {
+				bessNode, err = p.Ctx.Graph.NewNode(bessID, metadata)
+				if err != nil {
+					p.Ctx.Logger.Errorf("Failed to create bess node")
+					return
+				}
+			}
+			if !topology.HaveOwnershipLink(p.Ctx.Graph, p.Ctx.RootNode, bessNode) {
+				topology.AddOwnershipLink(p.Ctx.Graph, p.Ctx.RootNode, bessNode, nil)
+			}
+
+			// find ports to figure out pipeline roots
+			portMap := make(map[string]*graph.Node)
+			var bessPort *graph.Node
+			for _, port := range portRes.GetPorts() {
+				portID := graph.GenID(port.GetName())
+				bessPort = p.Ctx.Graph.GetNode(portID)
+				if bessPort == nil {
+					bessPort, err = p.Ctx.Graph.NewNode(portID, graph.Metadata{
+						"Name":    port.GetName(),
+						"MAC":     port.GetMacAddr(),
+						"Type":    port.GetDriver(),
+						"Backend": BessBackend,
+					})
+					if err != nil {
+						p.Ctx.Logger.Errorf("Failed to create bessPort: %v", bessPort)
```

Comment:
should we `continue` ? in order to avoid error later in AddOwnershipLink ?

Reply:
yes